h2o-core/src/main/java/water/parser/Parser.java

```diff
@@ -83,30 +89,64 @@ ParseWriter streamParseZip( final InputStream is, final StreamParseWriter dout,
     // All output into a fresh pile of NewChunks, one per column
     if (!_setup._parse_type.isParallelParseSupported) throw H2O.unimpl();
     StreamData din = new StreamData(is);
-    int cidx=0;
+    int cidx = 0;
     StreamParseWriter nextChunk = dout;
-    int zidx = bvs.read(null,0,0); // Back-channel read of chunk index
-    assert zidx==1;
-    while( is.available() > 0 ) {
-      int xidx = bvs.read(null,0,0); // Back-channel read of chunk index
-      if( xidx > zidx ) {  // Advanced chunk index of underlying ByteVec stream?
+    int zidx = bvs.read(null, 0, 0); // Back-channel read of chunk index
+    assert zidx == 1;
+//    int count = 0;
+
+    while (is.available() > 0) {
+      int xidx = bvs.read(null, 0, 0); // Back-channel read of chunk index
+      if (xidx > zidx) {  // Advanced chunk index of underlying ByteVec stream?
         zidx = xidx;       // Record advancing of chunk
         nextChunk.close(); // Match output chunks to input zipfile chunks
-        if( dout != nextChunk ) {
+        if (dout != nextChunk) {
           dout.reduce(nextChunk);
           if (_jobKey != null && _jobKey.get().stop_requested()) break;
         }
         nextChunk = nextChunk.nextChunk();
       }
+
+      if (cidx == 0)  {
+        // perform check header for each file again in order to determine if that file contains a header
+        // if column names are found in first row, set the _check_header field accordingly.
+        byte[] headerBytes = ZipUtil.unzipForHeader(din.getChunkData(cidx), this._setup._chunk_size);
+        ParserProvider pp = ParserService.INSTANCE.getByInfo(this._setup.getParseType());
+        ParseSetup ps = pp.guessSetup(null, headerBytes, this._setup._separator, this._setup.getColumnTypes().length,
```

Comment:
Can you use following instead here? (To be consistent with how this is done in multifiel parse task.)

" localSetup._check_header = localSetup.parser(_jobKey).fileHasHeader(ZipUtil.unzipBytes(zips,cpr, localSetup._chunk_size), localSetup);"